cobra/__init__.py

```diff
@@ -1158,17 +1160,52 @@ def initSocketBuilder(host,port):
         socket_builders[ (host,port) ] = builder
     return builder
 
-def startCobraServer(host="", port=COBRA_PORT):
+def startCobraServer(host="", port=COBRA_PORT, sslca=None, sslcrt=None, sslkey=None, msgpack=True):
+    # TODO: replace this with getCobraDaemon code?
     global daemon
     if daemon is None:
-        daemon = CobraDaemon(host,port)
+        daemon = CobraDaemon(host, port, sslca=sslca, sslcrt=sslcrt, sslkey=sslkey, msgpack=msgpack)
         daemon.fireThread()
     return daemon
 
-def runCobraServer(host='', port=COBRA_PORT):
-    daemon = CobraDaemon(host,port)
+daemon_threads = {}
+def runCobraServer(host='', port=COBRA_PORT, sslca=None, sslsrt=None, sslkey=None, msgpack=True):
+    global daemon_threads
+    if (host, port) in daemon_threads:
+        ######## REDO USING getCobraDaemon
+        daemon = daemon_threads.get(host, port)
+        logger.info("CobraDaemon already exists on port %d, joining to that thread.", port)
+        daemon.thr.join()
+
+    daemon = CobraDaemon(host, port, sslca=sslca, sslcrt=sslcrt, sslkey=sslkey, msgpack=msgpack)
     daemon.serve_forever()
 
+def registerCobraDaemon(host, port, daemon):
+    global daemon_threads
+    if (host, port) in daemon_threads:
+        raise Exception("Daemon already exists on host %r port %d.  Please deregister that port first." % (host, port))
+
+    daemon_threads[(host, port)] = daemon
+
+def deregisterCobraDaemon(host, port):
+    global daemon_threads
+    if port in daemon_threads:
+        daemon_threads.pop(port)
+
+    else:
+        logger.warning("Attempted to deregister a CobraDaemon on a port that isn't registered: %d", port)
+
+def getCobraDaemon(host="", port=COBRA_PORT, sslca=None, sslsrt=None, sslkey=None, msgpack=True):
+    global daemon_threads
```

Comment:
I have a lot of questions and concerns here.
* Why are we doubling down on this? I'm super not a fan of storing global state like this.
* In `registerCobraDaemon` you set the dictionary key via a tuple but retrieve/pop it via just the port. I think that's a bug?
* Who uses these?

Reply:
we already store a global state for one CobraDaemon.  this simply allows us to store and access multiples.

yes, that was a bug.  i fixed that bug and cleaned up the code a bit, and added some unittests for these helper functions.

Reply:
these are helper functions.  i use them.  they're made for third-party use.  the majority of Viv uses cobra for more specific purposes.

Reply:
i get that.  but we're already using a global for getCobraDaemon.  we're just changing it to be more robust. :)
it must be global, because we can only have one instance running on a given port.  period.  actually, across instances, in case we have multiples running in difference processes.  but that's tomorrow's problem.

Reply:
I know, but we're still doubling down on this odd global-state management pattern that feels like it's very much lacking in real fleshed out management/flexibility/error handling/support. It feels very much like a footgun we're walking into and a pattern we're going to regret supporting and that is screaming for something just more. I'd like to have a more robust API, maybe centered around something like a DaemonManager object that has a bit more ease of API addition/editing/etc that we can have people write to so our front facing API is a little friendlier and also easier for us to support/add to/discover/etc.

Reply:
Doubling Down on functionality which I believe should stay, regardless of the way we do it.
this is the best way that i can think of currently, without a *lot* of rework which would take a significant amount of planning and refactoring/engineering.

this PR involves an important bugfix, and some unittests, so i'd like to get it merged soon.

this approach is in line with what already exists, and doesn't make it harder to later make it work a different way (as long as we keep the functionality somehow).  can we merge now and come up with another solution later (since none seem to present themselves and we've gone round and round and consumed more than a little alcohol debating it ;)

Reply:
if turning this into a class with all the management functions that gets instantiated as the one and only global to track existing CobraDaemons... if that makes you happier, i can make that happen.